docs/src/main/sphinx/develop/table-functions.rst

```diff
@@ -0,0 +1,133 @@
+===============
+Table functions
+===============
+
+Table function declaration
+--------------------------
+
+Trino supports adding custom table functions. They are declared by connectors
+through implementing dedicated interfaces.
+
+To declare a table function, you need to subclass ``ConnectorTableFunction``.
```

Comment:
we probably need to add a whole bunch more info about the mechanics .. like separate module or fork of specific connector, pom.xml, dependencies?, what about package structure and so on..

Reply:
I added another sentence about passing the custom implementation through a Provider. It is all that the implementor must do. Also, this is a pretty common pattern of exposing things from connectors in Trino.

Reply:
I thought the user has to do the development .. and that is not the case it seems.. so it just goes into existing connectors .. we should mention that somehow

Reply:
When thinking of it, I can identify at least four categories of table functions, and they will be added in different ways:
1. functions that we will add to the existing connectors. The query pass-through capability is a good example, because it is specific to the connector: you pass different syntax to Elasticsearch than to Teradata.
2. functions that users add to their own connectors (forked). Generally, these are functions too controversial or too specific to be included in Trino
3. "global" functions, which will be built-in in Trino, but they don't belong to any specific connector. Example: the pivot capability. We will support them only when we have the dedicated Operator
4. I think that having table function support might result in creating a whole new kind of connectors. Currently, we have connectors over different data sources. Now that we support adding table functions, it makes sense to connect to an outer system for its _processing capabilities_ and not necessarily for the _data_. Such connector could expose table functions that give access to the outer system's abilities or resources, so that they can be used for processing data in Trino.

@mosabua could you help me with the wording? The problem is, the table functions can go both to the existing and new connectors. I can't predict what kind of use case will be dominant.